benchmarks: add store libary to comparison.

qnikst opened the pull request:
Store library was added to comparison.
Currently benchmarks for store are using generic instances,
so possibly using manual instances may improve results.

All benchmarks for store uses strict bytestrings,
this may be not fair but doesn't change results.

deserialiseNul benchmark is missing at the moment.

dcoutts commented:
Thanks!

dcoutts commented on bench/versus/Micro.hs:
```diff
@@ -60,6 +61,7 @@ benchmarks =
       , bench "aeson generic" (nf perfDecodeAesonGeneric tstdataA)
       , bench "aeson TH"      (nf perfDecodeAesonTH      tstdataA)
       , bench "read/show"     (nf perfDecodeReadShow     tstdataS)
+      , bench "read/show"     (nf perfDecodeReadShow     tstdataS)
```
Is this right? looks like a dupe

dcoutts commented on bench/versus/Micro.hs:
```diff
@@ -80,6 +82,7 @@ benchmarks =
     tstdataS = Micro.ReadShow.serialise tstdata
 --  tstdataM = PkgMsgpack.serialise tstdata
 --  tstdataN = Micro.NewMsgpack.serialise tstdata
+    tstdataR = Micro.Store.serialise tsdata
```
duplicating the var name?

dcoutts commented on bench/versus/Micro.hs:
```diff
@@ -98,6 +102,7 @@ benchmarks =
     perfDecodeAesonGeneric = Micro.PkgAesonGeneric.deserialise
     perfDecodeAesonTH      = Micro.PkgAesonTH.deserialise
     perfDecodeReadShow     = Micro.ReadShow.deserialise
+    perfDecodeReadShow     = Micro.PkgStore.deserialise
```
dupe?

dcoutts commented on bench/versus/Micro.hs:
```diff
@@ -88,6 +91,7 @@ benchmarks =
     perfEncodeAesonGeneric = BS.length . Micro.PkgAesonGeneric.serialise
     perfEncodeAesonTH      = BS.length . Micro.PkgAesonTH.serialise
     perfEncodeReadShow     = BS.length . Micro.ReadShow.serialise
+    perfEncodeReadShow     = BS.length . Micro.PkgStore.serialise
```
dupe?

dcoutts commented on bench/versus/Micro/PkgStore.hs:
```diff
@@ -0,0 +1,15 @@
+serialise pkgs = BS.fromStrict (Store.encode pkgs)
+
+deserialise :: BS.ByteString -> Tree
+deserialise = (\(Right x) -> x) . Store.decode . BS.toStrict
```
Mm, we should think about how to do fair comparison between libs when one of them forces the use of strict bytestrings.

dcoutts commented:
@thoughtpolice I don't understand how travis thinks this compiles. Is it compiling the benchmarks?

qnikst commented:
@dcoutts it should compile because it doesn't load problematic files. I'll address your comments and double check everything. Sorry for sending PR with such problems.

qnikst commented:
@dcoutts I have addressed all your comments. It appeared that they all were in a dead code, that have not been compiled when compiling benchmarks. I took responsibility for re-enabling that code in the same commit.

As for fair comparison, will it be ok  to use strict bytestrings to read file and convert that to lazy for libraries that want to have lazy input? Though I'm not sure about encoding.

dcoutts commented:
> As for fair comparison, will it be ok to use strict bytestrings to read file and convert that to lazy for libraries that want to have lazy input?

Yes, that gives the same best conditions to each lib.

For output/encoding, it's ok to have each one use it's native output style, ie no need to convert lazy to strict.

dcoutts commented:
@qnikst looks like the benchmark isn't building on travis.

```
bench/versus/Main.hs:10:18:
    Could not find module ‘Micro’
    Use -v to see a list of the files searched for.
```

qnikst commented:
Right, that file was not added to sdist as was missing in other modules, I'll test building tests after sdist locally.

qnikst commented:
@dcoutts 
1. I've moved to deserialisation to use  strict ByteStings.
2. Moved micro benchmarks versus store to `micro` benchmark
3. Remove unused files.

thoughtpolice commented:
Sorry this took so long - merged!
